src/Explorer/Controls/SmartUi/SmartUiComponent.tsx

```diff
@@ -101,42 +116,167 @@ export class SmartUiComponent extends React.Component<SmartUiComponentProps, Sma
+          numberInput.max = await (numberInput.max as numberPromise)();
+        }
+        if (numberInput.step instanceof Function) {
+          numberInput.step = await (numberInput.step as numberPromise)();
+        }
+        return numberInput;
+      }
+      case "boolean": {
+        const booleanInput = input as BooleanInput;
+        if (booleanInput.trueLabel instanceof Function) {
+          booleanInput.trueLabel = await (booleanInput.trueLabel as stringPromise)();
+        }
+        if (booleanInput.falseLabel instanceof Function) {
+          booleanInput.falseLabel = await (booleanInput.falseLabel as stringPromise)();
+        }
+        return booleanInput;
+      }
+      default: {
+        const enumInput = input as DropdownInput;
+        if (enumInput.choices instanceof Function) {
+          enumInput.choices = await (enumInput.choices as dropdownItemPromise)();
+        }
+        return enumInput;
+      }
+    }
+  };
+
+  private getDefaultValue = (input: AnyInput): InputType => {
+    switch (input.type) {
+      case "string": {
+        const stringInput = input as StringInput;
+        return stringInput.defaultValue ? (stringInput.defaultValue as string) : "";
+      }
+      case "number": {
+        return (input as NumberInput).defaultValue as number;
+      }
+      case "boolean": {
+        return (input as BooleanInput).defaultValue as boolean;
+      }
+      default: {
+        return (input as DropdownInput).defaultKey as string;
+      }
+    }
+  };
+
   private renderInfo(info: Info): JSX.Element {
     return (
       <MessageBar>
         {info.message}
-        <Link href={info.link.href} target="_blank">
-          {info.link.text}
-        </Link>
+        {info.link && (
+          <Link href={info.link.href} target="_blank">
+            {info.link.text}
+          </Link>
+        )}
       </MessageBar>
     );
   }
 
-  private onInputChange = (newValue: string | number | boolean, dataFieldName: string) => {
-    const { currentValues } = this.state;
-    currentValues.set(dataFieldName, newValue);
-    this.setState({ currentValues }, () => this.props.onChange(this.state.currentValues));
+  private onInputChange = (input: AnyInput, newValue: InputType) => {
+    if (input.onChange) {
+      const newValues = input.onChange(this.state.currentValues, newValue);
+      this.setState({ currentValues: newValues });
+    } else {
+      const dataFieldName = input.dataFieldName;
+      const { currentValues } = this.state;
+      currentValues.set(dataFieldName, newValue);
+      this.setState({ currentValues });
+    }
   };
 
   private renderStringInput(input: StringInput): JSX.Element {
+    const value = this.state.currentValues.get(input.dataFieldName) as string;
     return (
       <div className="stringInputContainer">
         <div>
           <TextField
             id={`${input.dataFieldName}-input`}
-            label={input.label}
+            label={input.label as string}
```

Comment:
Why casting here? Should avoid casting as much as possible

Reply:
removed all the casting